src/lib.rs

```diff
@@ -952,67 +966,103 @@ impl<'de, K: EnrKey> Deserialize<'de> for Enr<K> {
     }
 }
 
-impl<K: EnrKey> rlp::Encodable for Enr<K> {
-    fn rlp_append(&self, stream: &mut RlpStream) {
-        let include_signature = true;
-        self.append_rlp_content(stream, include_signature);
+impl<K: EnrKey> Encodable for Enr<K> {
+    fn encode(&self, out: &mut dyn bytes::BufMut) {
+        let mut list = Vec::<u8>::new();
+        self.signature.as_slice().encode(&mut list);
+        self.seq.encode(&mut list);
+        for (k, v) in &self.content {
+            k.as_slice().encode(&mut list);
+            list.extend_from_slice(v);
+        }
+        let header = Header {
+            list: true,
+            payload_length: list.len(),
+        };
+        header.encode(out);
+        out.put_slice(&list);
     }
 }
 
-impl<K: EnrKey> rlp::Decodable for Enr<K> {
-    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
-        if rlp.as_raw().len() > MAX_ENR_SIZE {
+impl<K: EnrKey> Decodable for Enr<K> {
+    fn decode(buf: &mut &[u8]) -> Result<Self, DecoderError> {
+        if buf.len() > MAX_ENR_SIZE {
             return Err(DecoderError::Custom("enr exceeds max size"));
         }
 
-        if !rlp.is_list() {
-            debug!("Failed to decode ENR. Not an RLP list: {}", rlp);
-            return Err(DecoderError::RlpExpectedToBeList);
+        let header = Header::decode(buf)?;
+
+        if !header.list {
+            return Err(DecoderError::Custom("Invalid format of header"));
         }
 
-        // verify there is no extra data
-        let payload_info = rlp.payload_info()?;
-        if rlp.as_raw().len() != payload_info.header_len + payload_info.value_len {
-            return Err(DecoderError::RlpInconsistentLengthAndData);
+        let payload = &mut &buf[..header.payload_length];
```

Comment:
this can panic if buf is of less size than `payload_length`. You need to check the size of the buffer first

Reply:
@armaganyildirak - I think these needs to be addressed. 

We should probably fuzz these changes as they are extensive and involve the decoder

Reply:
I think ```payload_lenght``` should be less size than buffer size but I will add the size check.

Reply:
oh yeah, i see. Nice